MAINT: adjustments to test_logistic.py::test_dtype_match

oleksandr-pavlyk opened the pull request:
When using Intel (R) DAAL to compute logistic loss and gradient with `solver='newton-cg'` the [test `test_dtype_match`](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/tests/test_logistic.py#L1363-1396) would [fail the last assertion](https://github.com/IntelPython/daal4py/issues/86), because coefficients computed in single and double precisions would only agree with relative tolerance of 3e-5, rather than 1e-6.

The change introduces use of `tol=solver_tol` in the logistic regression setup,  and derives absolute tolerance in parameter comparison from the data and the `solver_tol`.

The derivation starts with `dw ~= inv(H).grad`, where `grad` and `H` are the gradient and the Hessian at the solution, and `dw` is the distance to the true minimizer. The `solver_tol` bounds the norm of the gradient, and so one should use absolute tolerance of `norm(inv(H)) * solver_tol` when comparing  single and double precision solutions.

@ogrisel @jeremiedbb

oleksandr-pavlyk committed:
MAINT: adjustments to test_logistic.py::test_dtype_match

When using Intel (R) DAAL to compute logistic loss and gradient
with solver='newton-cg' the test test_dtype_match would fail the
last assertion, because coefficients computed in single and double
precisions would only agree with relative tolerance of 3e-5, rather
than 1e-6.

The change introduces solver_tol, and derives absolute tolerance
in parameter comparison from the data and the solver_tol

oleksandr-pavlyk commented:
@ogrisel @jeremiedbb ping

ogrisel commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1368,32 +1368,39 @@ def test_dtype_match(solver, multi_class):
+    #
+    # with  Z = np.hstack((np.ones((3,1)), np.array(X)))
+    # In [18]: np.linalg.norm(np.linalg.inv(0.25 * Z.T @ Z))
+    # Out[18]: 41.32740565066648
```
What is this supposed to explain?

ogrisel replied:
Ok so `0.25 * Z.T @ Z` is the Hessian for a model that outputs p(y_k|x) = 0.5 for each class for a binary classification problem. But at optimum, it's unlikely that p(y_k|x) is close to 1.

oleksandr-pavlyk replied:
I am afraid my mistake here is to not take into account the regularization term.

```
In [7]: np.linalg.norm(np.linalg.inv( np.diag([0, 2, 2]) + 0.25 * Z.T @ Z))
Out[7]: 1.7193336918135917
```

this is consistent with output of the following simulation:

```python
import sklearn, numpy as np, sklearn.linear_model
import sklearn.linear_model.logistic as logistic
from sklearn.preprocessing import LabelBinarizer

X = np.array([[-1, 0], [0, 1], [1, 1]], dtype=np.double)
Y1 = np.array([0, 1, 1], dtype='i')

Y_multi = LabelBinarizer().fit_transform(Y1)
Y_multi = np.hstack([1 - Y_multi, Y_multi])

lr = sklearn.linear_model.LogisticRegression(
        solver='lbfgs', multi_class='multinomial',
        random_state=42, tol=0., max_iter=1000, 
        fit_intercept=True)
lr.fit(X, Y1)

# Optimal value obtained by minimizing in extended precision and coercing
w_exact = np.array([0.47518386020515844474, 0.36183930896547519212, 0.19148890867094416443])

def grad(w_pos):
    w = np.hstack((-w_pos, w_pos))
    C = 1.0
    alpha = 1/C
    val, grad, p = logistic._multinomial_loss_grad(w, X, Y_multi, alpha, np.ones_like(Y1))
    return grad


w_pos_fitted = np.hstack((lr.coef_, lr.intercept_.reshape((-1,1)))).ravel()

def random_unit_vectors(dim, size=1):
    x = np.random.randn(size, dim)
    x /= np.linalg.norm(x, ord=2, axis=1, keepdims=True)
    return x

def norm_inf(u):
    return np.linalg.norm(u, ord=np.inf)

print(norm_inf(grad(w_exact)))    # 1.6653345369377348e-16 

solver_tol = 1e-5
print(np.max([
   norm_inf(grad(w_exact + solver_tol*ui)) for ui in random_unit_vectors(3, 10**4)
]))

# 1.695041656418761e-05 ~ 1.695 * solver_tol
```

which gives a close constant of `1.695 * solver_tol`.

 It is odd that norm of gradient on the solution is actually much higher than the value of `tol` provided to `lbfgs` solver:

```python
print(norm_inf(grad(w_pos_fitted))) # 2.88807176396233e-07
```

oleksandr-pavlyk committed:
Corrected computation of absolute tolerance

Added reference to the PR, which contains simulation code,
confirming theoretical estimate cited in the comments.

Also increase solver_tol to 5e-4, a little higher than the square root
of single precision epsilon:

```
In [3]: np.sqrt(np.finfo(np.float32).eps) * 10**5
Out[3]: 34.52669770922512
```

The previous value of 1e-5 may have been too tight for Newton CG code.

oleksandr-pavlyk commented:
All builds are green. I do not think the failure of codecov is meaningful, but I contend I may misunderstand its result.

agramfort reviewed (approved):
@jeremiedbb @ogrisel feel free to merge

ogrisel reviewed (approved):
This looks good to me as well. Thanks for the detailed comment.
